lib/wallaroo/core/sink/connector_sink/connector_sink.pony

```diff
@@ -542,148 +547,67 @@ actor ConnectorSink is Sink
-
-    let cpoint_id = ifdef "test_disconnect_at_5" then "5" else "" end
-    let drop_phase2_msg = try if _twopc.txn_id.split("=")(1)? == cpoint_id then true else false end else false end
-
-    _twopc.checkpoint_complete(this, drop_phase2_msg)
-
-    @ll(_twopc_debug, "2PC: DBGDBG: checkpoint_complete: commit, _twopc.last_offset %d _notify.twopc_txn_id_last_committed %s".cstring(), _twopc.last_offset, _notify.twopc_txn_id_last_committed_helper().cstring())
-
-    if _twopc.txn_id == "" then
-      // There are two reasons for this:
-      // 1. There is a legitimate bug.
-      // 2. We crashed and restarted, and we're still very early in
-      //    the restart process, e.g., during "INIT PHASE II".
-      //    Now we're told that the checkpoint is complete.  We didn't
-      //    really particpate in this checkpoint (because we had
-      //    crashed), but we can continue.
-      // We can't easily tell the difference between #1 and #2 because
-      // we would need to query LocalTopologyInitializer _initializer
-      // but can't.  I've ironed out most bugs in the 2PC impl, so
-      // let's assume that this isn't a fatal error.
-      @ll(_twopc_info, "checkpoint_complete() with empty _twopc.txn_id = %s.".cstring(), _twopc.txn_id.cstring())
-    else
-      if not _twopc.txn_id.contains("skip--.--CheckpointBarrierToken") then
-        _notify.twopc_txn_id_last_committed = _twopc.txn_id
-      end
-      @ll(_twopc_debug, "2PC: DBGDBG: twopc_txn_id_last_committed = %s.".cstring(), _notify.twopc_txn_id_last_committed_helper().cstring())
-    end
-    _twopc.reset_state()
-
-    _resume_processing_messages()
+    @ll(_twopc_debug, "2PC: Checkpoint complete %d _twopc.txn_id is %s".cstring(), checkpoint_id, _twopc.txn_id.cstring())
+    _cprb.checkpoint_complete(this, checkpoint_id)
+    @ll(_twopc_debug, "2PC: Checkpoint complete %d _twopc.txn_id is %s _cprb.name = %s".cstring(), checkpoint_id, _twopc.txn_id.cstring(), _cprb.name().cstring())
 
-    if drop_phase2_msg then
-      // Because we're using TCP, we get message loss only when
-      // the TCP connection is closed.  It doesn't matter why the
-      // connection is closed.  We have direct control over the
-      // timing here, so close it now.
-      _hard_close()
-      _schedule_reconnect()
-    end
-
-  fun ref _resume_processing_messages() =>
-    """
-    2nd-half logic for barrier_fully_acked().
-    """
-    _phase.resume_processing_messages()
-
-  fun ref resume_processing_messages_queued() =>
+  fun ref resume_processing_messages_queued(discard_app_msgs: Bool = false) =>
     let queued = _phase.queued()
     _use_normal_processor()
+    @ll(_conn_debug, "resume_processing_messages_queued: size = %lu".cstring(), queued.size())
     for q in queued.values() do
       match q
       | let qm: QueuedMessage =>
-        qm.process_message(this)
+        if not discard_app_msgs then
+          // We cannot use qm.process_message(this), because that
```

Comment:
Instead of this note, should we remove the process_message method from QueuedMessage to prevent this bug everywhere?  Or is there a case it's needed?

Reply:
`process_message()` is still in use by the other sinks ... at least until #3097 is fixed.